### PR TITLE
docs: better docs and types for the API

### PR DESCRIPTION
**Pull request by ozyx**
### Describe your changes:

- Update docs and types for just about all of Open MCT's API
- Export commonly used types from the top level of the typings files
  - This makes common types like `DomainObject`, `Identifier`, `ObjectPath`, etc. available at the root level so they can be used easily in JSDocs
- Removes the alias for `EventEmitter` so that `tsc` correctly generates types for classes that extend it

https://github.com/user-attachments/assets/f928a4d6-df64-4b66-afc6-c0cdbd92e5fb

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [ ] Is this a [notable change](../docs/src/process/release.md) that will require a special callout in the release notes? For example, will this break compatibility with existing APIs or projects that consume these plugins?

### Author Checklist

* [x] Changes address original issue?
* [ ] Tests included and/or updated with changes?
* [x] Has this been smoke tested?
* [x] Have you associated this PR with a `type:` label? Note: this is not necessarily the same as the original issue.
* [x] Have you associated a milestone with this PR? Note: leave blank if unsure.
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [ ] Code style and in-line documentation are appropriate?

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -10,7 +10,12 @@ const config = {
     serviceworker: true
   },
   globals: {
-    _: 'readonly'
+    _: 'readonly',
+    __webpack_public_path__: 'writeable',
```
Comment (scottbell) [inline .eslintrc.cjs:14]: nice. i just addded one of these to my branch, but this is more expansive

File: .webpack/webpack.common.mjs [hunk 2]
```diff
@@ -70,7 +70,6 @@ const config = {
       '@': path.join(projectRootDir, 'src'),
       legacyRegistry: path.join(projectRootDir, 'src/legacyRegistry'),
       csv: 'comma-separated-values',
-      EventEmitter: 'eventemitter3',
```
Comment (scottbell) [inline .webpack/webpack.common.mjs:73]: woah - big change

File: src/MCT.js [hunk 3]
```diff
@@ -73,37 +72,46 @@ import Browse from './ui/router/Browse.js';
...
+  /**
+   * Tracks current selection state of the application.
+   * @type {Selection}
+   */
+  selection;
   constructor() {
     super();
 
     this.buildInfo = {
+      // @ts-ignore
```
Comment (scottbell) [inline src/MCT.js:98]: is there a way we can add this to our ts config in some fashion? like we aded them to the .eslintrc?

File: .eslintrc.cjs (hunk 1 above)
Comment (scottbell) [reply .eslintrc.cjs:14]: apparently `writeable` vs `writable` is a big controversy:
https://github.com/eslint/eslint/pull/11338#discussion_r252495166

File: src/api/user/UserAPI.js [hunk 4]
```diff
@@ -20,35 +20,35 @@
...
-  constructor(openmct, config) {
+  constructor(openmct) {
     super();
 
     this.#openmct = openmct;
     this._provider = undefined;
 
     this.User = User;
-    this.status = new StatusAPI(this, openmct, config);
+    this.status = new StatusAPI(this, openmct);
```
Comment (scottbell) [inline src/api/user/UserAPI.js:45]: nice catch

File: src/ui/router/ApplicationRouter.js [hunk 5]
```diff
@@ -20,33 +20,36 @@
...
+   */
+
   constructor(openmct) {
     super();
 
     this.locationBar = new LocationBar();
     this.openmct = openmct;
     this.routes = [];
     this.started = false;
+    this.path = undefined;
```
Comment (scottbell) [inline src/ui/router/ApplicationRouter.js:52]: can we use null here instead? i think not as downstream we're checking only if it's `undefined`, but i always think it's weird to define something as `undefined`!

File: src/utils/clock/DefaultClock.js [hunk 6]
```diff
@@ -49,38 +45,53 @@ export default class DefaultClock extends EventEmitter {
...
     if (this.listeners(event).length === 0) {
       this.stop();
     }
 
-    return result;
+    return this;
+  }
+
+  stop() {
+    // Not implemented
```
Comment (scottbell) [inline src/utils/clock/DefaultClock.js:86]: should we throw an exception here if it's called?

File: src/utils/duration.js [hunk 7]
```diff
@@ -63,7 +63,21 @@ export function millisecondsToDHMS(numericDuration) {
...
+ * @param {number} value
+ * @param {Object} options
+ * @param {boolean | undefined} options.excludeMilliSeconds
+ * @param {boolean | undefined} options.useDayFormat
+ * @returns {string}
+ */
+export function getPreciseDuration(
+  value,
+  { excludeMilliSeconds, useDayFormat } = {
+    excludeMilliSeconds: undefined,
```
Comment (scottbell) [inline src/utils/duration.js:77]: again, seems weird to define something as undefined

File: tsconfig.json [hunk 8]
```diff
@@ -14,17 +14,7 @@
...
-    "outFile": "dist/types/index.d.ts",
-    "skipLibCheck": true,
-    "target": "ES2015",
-    "paths": {
-      // matches the alias in webpack config, so that types for those imports are visible.
-      "@/*": [
-        "src/*"
-      ]
-    }
+    "outFile": "dist/types/index.d.ts"
```
Comment (scottbell) [inline tsconfig.json:17]: look at all the deletion - wunderbar!

**Review by scottbell — COMMENTED** [review]
Looking good @ozyx! Had a few suggestions below. Also tested renderWhenVisible changes and they seemed to work nicely.

File: .webpack/webpack.common.mjs (hunk 2 above)
Comment (ozyx) [reply .webpack/webpack.common.mjs:73]: so i learned that `tsc` and webpack aliases don't play well together at all. After struggling to get `tsc` to recognize that many of our classes extend `EventEmitter` and have `.on()`, `.off()` methods, etc. I decided to just get rid of the alias altogether.

Comment (ozyx) [reply .webpack/webpack.common.mjs:73]: in general i think we have a lot of unnecessary aliases

File: src/ui/router/ApplicationRouter.js (hunk 5 above)
Comment (ozyx) [reply src/ui/router/ApplicationRouter.js:52]: yes yes yes

File: src/utils/visibility/VisibilityObserver.js [hunk 9]
```diff
@@ -50,12 +67,16 @@ export default class VisibilityObserver {
     this.renderWhenVisible = this.renderWhenVisible.bind(this);
   }
 
+  /**
+   * @returns {boolean}
+   */
   #inOverlay() {
-    return this.#element.closest('.js-overlay') !== null;
+    return Boolean(this.#element?.closest('.js-overlay'));
```
[the remark this answers is not in the document]
Comment (ozyx) [reply src/utils/visibility/VisibilityObserver.js:74]: I'll revert

File: src/MCT.js (hunk 3 above)
Comment (ozyx) [reply src/MCT.js:98]: ooooo i think these actually snuck in when i was doing some experimenting. we have ts-check disabled, but i enabled it for a while to guide my type updates

File: tsconfig.json (hunk 8 above)
Comment (ozyx) [reply tsconfig.json:17]: right! we're not transpiling typescript, so i was able to remove all of the options that are solely meant for that

**Review by scottbell — APPROVED** [review]
Good stuff!

**Pull request merged**